lib/vnet/admin_process_windows.go

```diff
@@ -150,7 +164,189 @@ func newWindowsNetworkStackConfig(tun tunDevice, clt *clientApplicationServiceCl
 	}, nil
 }
 
-func authenticateUserProcess(ctx context.Context, clt *clientApplicationServiceClient) error {
-	// TODO(nklaassen): implement process authentication.
+func authenticateUserProcess(ctx context.Context, clt *clientApplicationServiceClient, userSID string) error {
+	pipe, err := createNamedPipe(ctx, userSID)
+	if err != nil {
+		return trace.Wrap(err, "creating named pipe")
+	}
+	defer pipe.Close()
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		if err := clt.AuthenticateProcess(ctx, pipe.name); err != nil {
+			return trace.Wrap(err, "authenticating user process")
+		}
+		return nil
+	})
+	g.Go(func() error {
+		if err := validateClientExe(ctx, pipe); err != nil {
+			return trace.Wrap(err, "validating user process exe")
+		}
+		return nil
+	})
+	if err := g.Wait(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func validateClientExe(ctx context.Context, p *winpipe) error {
+	thisExePath, err := os.Executable()
+	if err != nil {
+		return trace.Wrap(err, "getting executable path for this service")
+	}
+	clientExePath, err := p.waitForClient(ctx)
+	if err != nil {
+		return trace.Wrap(err, "waiting for client to connect to named pipe")
+	}
+	log.DebugContext(ctx, "Got pipe connection from client", "exe", clientExePath)
+	if err := compareFiles(thisExePath, clientExePath); err != nil {
+		return trace.AccessDenied(
+			"remote process is not running the same executable as this service, remote process exe: %s, this process exe: %s",
+			clientExePath, thisExePath)
+	}
+	return nil
+}
+
+func compareFiles(p1, p2 string) error {
+	h1, err := hashFile(p1)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	h2, err := hashFile(p2)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !bytes.Equal(h1, h2) {
+		return trace.CompareFailed("files %s and %s are not equal", p1, p2)
+	}
+	return nil
+}
+
+func hashFile(path string) ([]byte, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, trace.Wrap(err, "opening %s", path)
+	}
+	defer f.Close()
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return nil, trace.Wrap(err, "hashing %s", path)
+	}
+	return h.Sum(nil), nil
+}
+
+type winpipe struct {
+	pipeHandle  windows.Handle
+	eventHandle windows.Handle
+	name        string
+}
+
+// createNamedPipe creates a new Windows named pipe, connects to it as a server
+// and sets up the pipe to receive client connections.
+func createNamedPipe(ctx context.Context, userSID string) (*winpipe, error) {
+	pipeName := `\\.\pipe\` + uuid.NewString()
+	pipePath, err := syscall.UTF16PtrFromString(pipeName)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting string to UTF16")
+	}
+	// Allow GA (Generic All) pipe access to the user.
+	sdString := fmt.Sprintf("D:P(A;;GA;;;%s)", userSID)
```

Comment:
Is there some API Reference or documentation we can link to here for the format of this string? It's a little difficult to understand what this means.

Reply:
yeah added a link